markup: add --citeproc to pandoc converter

shoeffner opened the pull request:
Adds the citeproc filter to the pandoc converter.

There are several PRs for it this feature already. However, I think
simply adding `--citeproc` is the cleanest way to enable this feature,
with the option to flesh it out later, e.g., in #7529.

Some PRs and issues attempt adding more config options to Hugo which
indirectly configure pandoc, but I think simply configuring Pandoc via
Pandoc itself is simpler, as it is already possible with two YAML
blocks -- one for Hugo, and one for Pandoc:

    ---
    title: This is the Hugo YAML block
    ---
    ---
    bibliography: assets/pandoc-yaml-block-bibliography.bib
    ...
    Document content with @citation!

There are other useful options, e.g., #4800 attempts to use `nocite`,
which works out of the box with this PR:

    ---
    title: This is the Hugo YAML block
    ---
    ---
    bibliography: assets/pandoc-yaml-block-bibliography.bib
    nocite: |
      @*
    ...
    Document content with no citations but a full bibliography:

    ## Bibliography

Other useful options are `csl: ...` and `link-citations: true`, which
set the path to a custom CSL file and create HTML links between the
references and the bibliography.

The following issues and PRs are related:

- Add support for parsing citations and Jupyter notebooks via Pandoc and/or Goldmark extension #6101
  Bundles multiple requests, this PR tackles citation parsing.

- WIP: Bibliography with Pandoc #4800
  Passes the frontmatter to Pandoc and still uses
  `--filter pandoc-citeproc` instead of `--citeproc`.
- Allow configuring Pandoc #7529
  That PR is much more extensive and might eventually supersede this PR,
  but I think --bibliography and --citeproc should be independent
  options (--bibliography should be optional and citeproc can always be
  specified).
- Pandoc - allow citeproc extension to be invoked, with bibliography. #8610
  Similar to #7529, #8610 adds a new config option to Hugo.
  I think passing --citeproc and letting the users decide on the
  metadata they want to pass to pandoc is better, albeit uglier.

Note:
This PR also adds a tiny little bit of unrelated documentation to the external helpers by pointing out MathJax superficially. This certainly needs improvements, but is out of the scope for this PR.

shoeffner committed:
markup: add --citeproc to pandoc converter

Adds the citeproc filter to the pandoc converter.

There are several PRs for it this feature already. However, I think
simply adding `--citeproc` is the cleanest way to enable this feature,
with the option to flesh it out later, e.g., in #7529.

Some PRs and issues attempt adding more config options to Hugo which
indirectly configure pandoc, but I think simply configuring Pandoc via
Pandoc itself is simpler, as it is already possible with two YAML
blocks -- one for Hugo, and one for Pandoc:

    ---
    title: This is the Hugo YAML block
    ---
    ---
    bibliography: assets/pandoc-yaml-block-bibliography.bib
    ...
    Document content with @citation!

There are other useful options, e.g., #4800 attempts to use `nocite`,
which works out of the box with this PR:

    ---
    title: This is the Hugo YAML block
    ---
    ---
    bibliography: assets/pandoc-yaml-block-bibliography.bib
    nocite: |
      @*
    ...
    Document content with no citations but a full bibliography:

    ## Bibliography

Other useful options are `csl: ...` and `link-citations: true`, which
set the path to a custom CSL file and create HTML links between the
references and the bibliography.

The following issues and PRs are related:

- Add support for parsing citations and Jupyter notebooks via Pandoc and/or Goldmark extension #6101
  Bundles multiple requests, this PR tackles citation parsing.

- WIP: Bibliography with Pandoc #4800
  Passes the frontmatter to Pandoc and still uses
  `--filter pandoc-citeproc` instead of `--citeproc`.
- Allow configuring Pandoc #7529
  That PR is much more extensive and might eventually supersede this PR,
  but I think --bibliography and --citeproc should be independent
  options (--bibliography should be optional and citeproc can always be
  specified).
- Pandoc - allow citeproc extension to be invoked, with bibliography. #8610
  Similar to #7529, #8610 adds a new config option to Hugo.
  I think passing --citeproc and letting the users decide on the
  metadata they want to pass to pandoc is better, albeit uglier.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gohugoio/hugo?pullRequest=9953) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gohugoio/hugo?pullRequest=9953) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gohugoio/hugo?pullRequest=9953) it.</sub>

bep commented:
@shoeffner thanks for this. This PR is certainly easier to reason about (because of its size) than some of the others I've seen.

That said, the test failure on Linux shows that this needs some rethinking, as that flag is not available in the version we test on, fetched via:

```
sudo apt-get install -y pandoc
```

I assume this flag was added recently? I'm not sure how to handle this.

shoeffner commented:
The Linux build seems to use pandoc 2.5, but --citeproc is only supported > 2.11 (https://pandoc.org/releases.html#pandoc-2.11-2020-10-11 – on macOS, 2.18 is installed, so I didn't notice) – before `--filter pandoc-citeproc` was the way to go.

If I remember correctly, pandoc-citeproc had to be installed manually, but I will check it out in a container and see what I can do!

shoeffner commented:
choco and brew seem to install 2.18, so Windows and macOS should not need any specific handling.

On Linux we have everything from 2.5 (Ubuntu 20.04, e.g. GH actions) to 2.17 (Arch Linux); the most common version seems to be 2.9 (Debian stable, Ubuntu 22.04, and other distros), here `pandoc-citeproc` must be installed additionally.

This makes the pull request much more complicated than what I was aiming for, so feel free to close it again. My current approach is now:

```
if pandoc >= 2.11
    add --citeproc
else if pandoc-citeproc available
    add --filter pandoc-citeproc
otherwise
    keep the old behavior
```

If the citations are not supported (due to the version or pandoc-citeproc dependency), SupportsCitations is false and the corresponding tests are skipped. I also added a remark to the docs.
However, I was unable to test all configurations yet (i.e., Linux with newer pandoc, older pandoc, with and without pandoc-citeproc; nor did I test windows), thus I update the PR to "WIP".

An alternative could be to check for the version and ignore pandoc-citeproc. This would make it much simpler and whoever wants to use citeproc might also want to use other pandoc features anyways. But it puts some burden on the users who have to install a newer pandoc version.

What do you think?

bep commented on markup/pandoc/convert.go:
```diff
@@ -77,6 +80,64 @@ func getPandocBinaryName() string {
+}
+
+func getCiteprocArgs(cfg converter.ProviderConfig) []string {
+	pandocVersion, err := getPandocVersion(cfg)
```
I think it's safe to assume that the pandoc version can be cached (e.g. via a `sync.Once`). Shelling out to `os/exec` isn't cheap.

shoeffner replied:
Done!

bep commented on markup/pandoc/convert_test.go:
```diff
@@ -25,18 +25,114 @@ import (
 )
 
-func TestConvert(t *testing.T) {
+func SetupTestConverter(t *testing.T) (*qt.C, converter.Converter, converter.ProviderConfig) {
```
This can be unexported?

shoeffner replied:
I didn't know that concept, as I haven't done much programming in Go, thank you very much!

bep commented on markup/pandoc/convert.go:
```diff
@@ -77,6 +80,64 @@ func getPandocBinaryName() string {
+	return version, nil
+}
+
+func SupportsCitations(cfg converter.ProviderConfig) bool {
```
Can be unexported?

shoeffner commented:
I rebased the commits and squashed them.
Unless there is anything else or the CI fails again, the PR is for the moment done, so I will also remove the WIP label.

shoeffner commented:
I updated the PR again, I made a mistake (`err != nil` instead of `err == nil`)which caused the tests to skip and never activated pandoc. It should now work as expected.

shoeffner commented:
Rebased on master.

bep commented on markup/pandoc/convert.go:
```diff
@@ -77,6 +85,51 @@ func getPandocBinaryName() string {
+
+// getPandocVersion parses the pandoc version output
+func getPandocVersion(cfg converter.ProviderConfig) (string, error) {
+	var version string
```
You need to pull the `version` out of the function; this will fail he second time `getPandocVersion` gets called.

shoeffner replied:
Done, thanks.

shoeffner replied:
I also added a unit test which calls it twice to detect this issue.

bep commented on markup/pandoc/convert.go:
```diff
@@ -77,6 +85,51 @@ func getPandocBinaryName() string {
+	var err error
+
+	versionOnce.Do(func() {
+		argsv := collections.StringSliceToInterfaceSlice([]string{"--version"})
```
The above works, but it doesn't make any sense when you can just do `argsv := []string{"--version"}`

shoeffner replied:
I tried, but it does not work that way:

```go
argsv := []string{"--version"}

var out bytes.Buffer
argsv = append(argsv, hexec.WithStdout(&out))
```
does not compile:
```
markup/pandoc/convert.go:98:25: cannot use hexec.WithStdout(&out) (value of type func(c *hexec.commandeer)) as type string in argument to append
```
It works with StringSliceToInterfaceSlice, but don't ask me why – I adapted the whole function from markup/internal/external.go:
https://github.com/gohugoio/hugo/blob/8e2fd55923a4da38eb2ddda51dc1b96943be0c56/markup/internal/external.go#L23

shoeffner replied:
I assume that the `InterfaceSlice` is compatible to contain not only strings but also `func(*hexec.comandeer)`, so a conversion probably does make sense here. But again, I am new to Go, so I might be wrong.

bep replied:
Yea, right, it's a slice of `any`, you need to do

```go
argsv := []any{"--version"}
```

shoeffner commented:
Sorry for the late reaction. I fixed it to use `argsv := []any{"--version"}` instead of `collections.StringSliceToInterfaceSlice...`.

shoeffner commented:
Rebased on master.

Setr43 commented:
Вітаємо в буфері обміну Gboard. Скопійований текст зберігатиметься тут.

shoeffner commented:
Okay, comparing strings 2.5 and 2.11 was a stupid idea... Now it has a proper type and a function to compare two instances.

https://github.com/gohugoio/hugo/commit/3c5e31fd17898637f96e99900613ba9658a96fcf#diff-f662bf93836a7230b77cdb532095a04622eb9eda3054124e8ba9399786870efaR88-R94 

Additionally, I included some test cases for those comparisons. I hope it will work now ;-)

shoeffner commented:
Unsupported pandoc should now cause the tests to SKIP instead of causing a panic.

shoeffner commented:
Rebased on the current tip of the master branch.

jacobmerson commented:
@shoeffner thanks for putting this PR together. It is quite useful. Do you have a sense if this, or something like it will end up getting merged? Being able to put citations in text is quite important.

shoeffner commented:
Thanks for getting back to this, I rebased the changes on the master.

I don't know whether this will be merged and if the pipelines still work, nor about the progress on alternative approaches, as I haven't followed the current developments much. Maybe @bep knows more about it.

shoeffner commented:
Is this still relevant to Hugo? I have to admit that I don't even use it right now, so I haven't followed up anymore.
But if it's still useful, I am happy to rebase/update/work on the PR – otherwise, feel free to close it :)

shoeffner commented:
The docs moved around quite a bit over the last 1.5 years, so I moved the section about citations to its own markdown file during the rebase. Please let me know if I should change that (or, for example, the weight) – or simply change it.

I also found out that the test setup (or provider config) seems to have changed, I am currently working on updating that.

shoeffner commented:
I found the issue in the test setup (the ProviderConfig was not passed correctly, and I seem to have written that bug the last time I rebased in August '23, whoops).

There's also been some output format changes in Pandoc, so I will relax the tests a bit and make sure they only check for "contains" on the references, not the exact markup (as that will be different between Pandoc versions).

shoeffner commented:
Done, the tests no longer rely on a specific output format and only check if the author last names, the year, and a word from the title are included in the documents – or not included if no citations was there.

shoeffner commented:
I received an email [that the tests have failed](https://github.com/gohugoio/hugo/actions/runs/13347029218/job/37279359322), but I think that's what I already fixed in the latest commit.

alametti commented:
Just tested this rebase and it works fine; I built with `$ CGO_ENABLED=1 go build -tags extended .` after extracting.

### My current workflow with regular extended Hugo:

- Unprocessed markdown content goes in `${DRAFT_DIR}`
- CI pipeline preprocesses with `pandoc` and some ugly kludges:  
```sh
touch .dummy;
ls -1 ${DRAFT_DIR}/*.md \
| xargs basename \
| xargs -i pandoc-${PANDOC_VERSION}/bin/pandoc \
-C -t gfm --resource-path=${DRAFT_DIR} -H .dummy \
${DRAFT_DIR}/{} -o ${POST_DIR}/{}
```
- Build site with `hugo`

### Workflow with this PR:

- Markdown content lives in `content/`
- Add `markup: pandoc` to front matter
- Whitelist `pandoc` executable in `hugo.yaml`
- Build site with `hugo`

Much easier and no preprocessing needed!

jmooring commented:
@shoeffner I'm not in a position to judge the value of this addition, but I do have a question. Instead of all of the version checking code, can you just test the exit code?

```sh
pandoc --citeproc --dump-args
echo $? # 0 (success)

pandoc --foo --dump-args
echo $? # 6 (failure)
```

shoeffner commented:
> can you just test the exit code

Sounds plausible, I'll give it a try!


> Add `markup: pandoc` to front matter

Hm, should I add that to the documentation about citations I added? I think
the test examples work without that.

Thanks for trying it out!


On Fri, Feb 21, 2025, 05:26 Joe Mooring ***@***.***> wrote:

> @shoeffner <https://github.com/shoeffner> I'm not in a position to judge
> the value of this addition, but I do have a question. Instead of all of the
> version checking code, can you just test the exit code?
>
> pandoc --citeproc --dump-argsecho $? # 0 (success)
>
> pandoc --foo --dump-argsecho $? # 6 (failure)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gohugoio/hugo/pull/9953#issuecomment-2673381298>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOAOD4DRICQM3JUYNGDAD32Q2TI5AVCNFSM6AAAAABV7M5RZ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZTGM4DCMRZHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: jmooring]*jmooring* left a comment (gohugoio/hugo#9953)
> <https://github.com/gohugoio/hugo/pull/9953#issuecomment-2673381298>
>
> @shoeffner <https://github.com/shoeffner> I'm not in a position to judge
> the value of this addition, but I do have a question. Instead of all of the
> version checking code, can you just test the exit code?
>
> pandoc --citeproc --dump-argsecho $? # 0 (success)
>
> pandoc --foo --dump-argsecho $? # 6 (failure)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gohugoio/hugo/pull/9953#issuecomment-2673381298>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOAOD4DRICQM3JUYNGDAD32Q2TI5AVCNFSM6AAAAABV7M5RZ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZTGM4DCMRZHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

alametti commented:
> Add `markup: pandoc` to front matter

To clarify, this is because I did not set `pandoc` as the global markup engine, only for select content.

shoeffner commented:
Thanks, that makes sense!

On Fri, Feb 21, 2025, 15:02 alametti ***@***.***> wrote:

> Add markup: pandoc to front matter
>
> To clarify, this is because I did not set pandoc as the global markup
> engine, only for select content.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gohugoio/hugo/pull/9953#issuecomment-2674631796>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOAOD4QJBJBTDORBBXNEU32Q4WXNAVCNFSM6AAAAABV7M5RZ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZUGYZTCNZZGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: alametti]*alametti* left a comment (gohugoio/hugo#9953)
> <https://github.com/gohugoio/hugo/pull/9953#issuecomment-2674631796>
>
> Add markup: pandoc to front matter
>
> To clarify, this is because I did not set pandoc as the global markup
> engine, only for select content.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gohugoio/hugo/pull/9953#issuecomment-2674631796>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOAOD4QJBJBTDORBBXNEU32Q4WXNAVCNFSM6AAAAABV7M5RZ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZUGYZTCNZZGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jmooring commented:
> To clarify, this is because I did not set pandoc as the global markup engine, only for select content.

Why not  change the file extension to `.pdc` or `.pandoc`?

alametti commented:
> Why not change the file extension to `.pdc` or `.pandoc`?

Because I didn't know I could do that. Thanks for the tip - I'm not a hugo expert by any means!

patch-work commented:
> * Build site with `hugo`
> 
> ### Workflow with this PR:
> * Markdown content lives in `content/`
> * Add `markup: pandoc` to front matter
> * Whitelist `pandoc` executable in `hugo.yaml`
> * Build site with `hugo`
> 
> Much easier and no preprocessing needed!

Is it available on hugo 0.145.0?

shoeffner commented:
I changed the citeproc detection to use `--dump-args`, as suggested by @jmooring . It seems to work for me, so feel free to test it. I also rebased on the latest master version again.

shoeffner commented:
Hm, it seems the tests were unable to pull some data from X.com - I don't think that's related to my PR, though. Or am I mistaken?
The tests I wrote seem to have worked (markup/pandoc/convert_test.go).

bep commented:
>Hm, it seems the tests were unable to pull some data from X.com - I don't think that's related to my PR, though. Or am I mistaken?

The X.com API has been unstable lately.

jmooring commented:
You' re logging _and_ returning errors, resulting in repetitive console output:

```text
Start building sites … 
hugo v0.146.0-DEV-038007c117257f13acde3d2d230003bfd8161b9b+extended+withdeploy linux/amd64 BuildDate=2025-03-18T17:16:55Z VendorInfo=jmooring

ERROR Could not call pandoc: access denied: "pandoc" is not whitelisted in policy "security.exec.allow"; the current security configuration is:

[security]
  enableInlineShortcodes = false

  [security.exec]
    allow = ['^((dart-)?sass|git|go|npx|postcss|tailwindcss|gnome-text-editor|nano|asciidoctor|babel|xpandoc|rst2html)$']
    osEnv = ['(?i)^((HTTPS?|NO)_PROXY|PATH(EXT)?|APPDATA|TE?MP|TERM|GO\w+|(XDG_CONFIG_)?HOME|USERPROFILE|SSH_AUTH_SOCK|DISPLAY|LANG|SYSTEMDRIVE)$']

  [security.funcs]
    getenv = ['^HUGO_', '^CI$']

  [security.http]
    methods = ['(?i)GET|POST']
    urls = ['.*']
Total in 14 ms
Error: error building site: render: failed to render pages: render of "/home/jmooring/code/hugo-testing/content/_index.pdc" failed: "/home/jmooring/code/hugo-testing/layouts/_default/home.html:11:5": execute of template failed: template: _default/home.html:11:5: executing "main" at <.Content>: error calling Content: "/home/jmooring/code/hugo-testing/content/_index.pdc:1:1": access denied: "pandoc" is not whitelisted in policy "security.exec.allow"; the current security configuration is:

[security]
  enableInlineShortcodes = false

  [security.exec]
    allow = ['^((dart-)?sass|git|go|npx|postcss|tailwindcss|gnome-text-editor|nano|asciidoctor|babel|xpandoc|rst2html)$']
    osEnv = ['(?i)^((HTTPS?|NO)_PROXY|PATH(EXT)?|APPDATA|TE?MP|TERM|GO\w+|(XDG_CONFIG_)?HOME|USERPROFILE|SSH_AUTH_SOCK|DISPLAY|LANG|SYSTEMDRIVE)$']

  [security.funcs]
    getenv = ['^HUGO_', '^CI$']

  [security.http]
    methods = ['(?i)GET|POST']
    urls = ['.*']
```

shoeffner commented:
I removed the two log lines, is it better this way? Otherwise, can you share the example you are trying so that I can reproduce the behavior?

shoeffner commented:
@jmooring, I only ever run the tests with `mage -v check` and they do not produce any meaningful output as they suppress it all.

I don't know what example project you use to test the PR, even though it was clear to me that you removed pandoc from the allow list.
So please excuse that I asked whether it was better; what I wanted to ask is whether this was closer to your preferred behavior, as I was trying to incorporate your feedback. I did not mean to offend and I am sorry if I did.

Of course, I understand that I am not the ideal contributor to work on this because I haven't worked with Hugo since shortly after I originally drafted this PR in 2022, as such, I am not even using this feature myself anymore. I am trying my best to get a feel for the project and improve as I get suggestions, be it by you, @bep - who reviewed the first draft and helped me over a few Go hurdles -, or others in this thread.

If you feel I am not confident enough on contributing this, I will of course step back from the PR, you can close it, and/or someone else can pick it up.

foxyseta commented:
I am personally working with Hugo and `pandoc --citeproc` right now, if you need somebody to pick this up.

shoeffner commented:
Sure, I gave you access to my fork so you can work on this PR directly – but feel free to create your own if you like :)

Qodo-Free-For-OSS commented:
Hi, getPandocSupportsCiteproc shadows the outer `err` with `cmd, err := ...`, so the function always returns a nil error even when pandoc execution fails. This silently disables citeproc on failure and prevents callers/tests from detecting the real cause (e.g. exec policy denial or unsupported flag).

**Severity:** action required | **Category:** correctness

**How to fix:** Fix err shadowing/propagation

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> `getPandocSupportsCiteproc` declares an outer `err` but then shadows it inside the `sync.Once` closure (`cmd, err := ...`). As a result, the function always returns a nil error even when probing pandoc fails.
> 
> ### Issue Context
> The return signature includes `error`, and `supportsCitations` relies on `err == nil` to decide whether to add `--citeproc`. With the current bug, failures are indistinguishable from success from an error-reporting perspective.
> 
> ### Fix Focus Areas
> - markup/pandoc/convert.go[88-111]
> - markup/pandoc/convert_test.go[67-75]
> 
> ### What to change
> - Avoid shadowing by using `cmd, err = ...` (declare `var cmd hexec.Runner` or similar outside the closure) or rename the inner error variable and assign it to the outer one.
> - Ensure errors from `cfg.Exec.New(...)` and `cmd.Run()` are returned to the caller.
> - Update the test to assert the corrected behavior (e.g., if the probe fails, `err` should be non-nil; if it succeeds, `err` should be nil).

---
*Found by [Qodo](https://github.com/marketplace/qodo-merge-pro-for-open-source) code review. FYI, Qodo is free for open-source.*

shoeffner commented:
@Qodo-Free-For-OSS feel free to pick up the code in the PR and create your own. I am no longer working on this and will close it now.

Dear maintainers, feel free to take the code if there is any interest :)
